### DISCUSSION TITLE
avoid radius estimation when using "delaunay_triangulation"

### DISCUSSION THREAD

**Discussion by domagojhack** [raphaelquast/EOmaps#88] [Q&A] [answered]
I am successfully embedding EOmaps into my app. However, I have a slight "hiccup". My App will work with a lot of raster data.
However, each time plot is redrawn I see that EOmaps conduct something that is called Radius estimation. (It takes about a second or two per raster layer). As I am working with data that will have more and less the same resolution, extent etc. 
It looks like this `Radius estimation` could be circumvented using some precalculated `radius` value.

I tried to look in the source but could not trace the radius argument, only the estimator.

Here is a minimum code example:

```
import geopandas as gpd
from eomaps import Maps
import matplotlib
%matplotlib widget

m = Maps(layer="DEM")

data = m.read_file.GeoTIFF(
    "base_layers/dem/myDEM.tif")
m.set_shape.delaunay_triangulation()
m.set_data(**data)
m.plot_map(cmap="gist_earth")
```

Output I am getting:

```
EOmaps: estimating radius...
EOmaps: radius: 12.5000
```

How can I set `radius` so I can circumvent radius estimation?

Cheers

**Comment by raphaelquast** [comment, marked as answer]
Hey, sounds awesome! I'd love to get a look at it!

For delaunay-triangulations the radius-estimation is used differently than for other shapes...
The radius is only used as a "mask_radius" to mask the obtained triangulation 
(to avoid drawing very large triangles that connect far-away points).
If the mask-radius is not provided explicitly EOmaps tries to come up with a suitable guess based on the pixel-spacing (hence the radius-estimation).

To avoid the estimation in the current release, simply use `m.set_shape.delaunay_triangulation(mask_radius=np.inf)` (or use any number appropriate to your data). Not sure though if the radius-estimation is the slow part here... I'd expect the evaluation of the triangulation to take up most of the time... 

I'm actually in the process of improving radius-estimation a little bit (see  #87)...
In EOmaps >= v4.2.2  `m.set_shape.delaunay_triangulation(masked=False)` will also avoid the associated  radius estimation step.

**Comment by domagojhack** [comment]
Here is a sneak peek!

![sneak peek](https://user-images.githubusercontent.com/2804094/172607728-df305e3d-b945-414c-9a90-08693d72c7f0.png)

Great! Yeah I've got just a light speedup from `m.set_shape.delaunay_triangulation(mask_radius=np.inf)` probs only from the lack of print statements. Thanks for the info!

**Reply by raphaelquast** [reply]
Wow, that looks really nice! I didn't look a lot into embedding EOmaps in another gui so far but I'll definitely have to get a closer look on `wxpython` if I find the time for it 😄 (I especially like the fact that it just wraps system-ui elements... gets you started a lot quicker I guess)

- If you're datasets are large and speed is of concern, shapes such as "raster", "shade_raster" or "shade_points" should be a lot faster than Delaunay triangulations.
- ... and just to put it here as well.. If you intend to distribute the app on a broader scale using OSM-tiles, make sure to comply to their license!   (OSM is free but heavy server-bandwidth usage is not...   https://operations.osmfoundation.org/policies/tiles/ )


... also If by any chance you'd be willing to share some of your insights on the wxpython-embedding, I'd very much appreciate a section in the documentation explaining the basics on how to "Embed EOmaps in a wxpython gui" (maybe even with a minimal example...).

**Reply by domagojhack** [reply]
I don't use OSM-tiles per se, I am using offline OSM data I downloaded and extracted from daily-updated OSM files from [http://download.geofabrik.de/](http://download.geofabrik.de/). It is much faster and does not get affected by server-bandwidth limitations while being enough for my use case. (I plan, in the future, to host my own osm tile server for apps). It is a bit tedious to prepare data in a format that can be read using Geopandas but it works.

I am more than willing, I already started to work on the documents about embedding in wxpython and I am also researching possible embedding in [Kivy](https://kivy.org/#home) framework as well.

**Reply by raphaelquast** [reply]
OK, that's of course the optimal way to do it 🙂   (not tested but it seems they offer shapefiles as well  which `gpd.read_file()` should be able to read right away... (however for large amounts of polygons it tends to get quite slow)

> I am more than willing, I already started to work on the documents about embedding in wxpython and I am also researching possible embedding in [Kivy](https://kivy.org/#home) framework as well.

Nice, really looking forward to this! Let me know if you need any help!
Kivy sure looks interesting as well!

**Reply by domagojhack** [reply]
They provide vector data indeed but in form of binary `*.osm.pbf` data format, not  ESRI shapefiles. I managed to extract and transform data for my ROI, into something smaller, manageable, and readable with Geopandas. So the polygon number is not too big.
